File: JNSocialDownload/JNSocialDownload.m
//
//  SocialDownload.m
//
//
//  Created by Joao Nunes on 31/10/14.
//
//

#import "JNSocialDownload.h"

@import Social;
@import Accounts;



#pragma mark - RequestConfiguration

@interface RequestConfiguration : NSObject

@property (nonatomic) JNSocialDownloadNetwork network;

@property (nonatomic, readonly, copy) SocialDownloadUserInfoBlock infoBlock;
@property (nonatomic, readonly, copy) SocialDownloadImageBlock imageBlock;


+ (instancetype)requestConfigurationWithInfoBlock:(SocialDownloadUserInfoBlock)block andNetwork:(JNSocialDownloadNetwork)network;
+ (instancetype)requestConfigurationWithImageBlock:(SocialDownloadImageBlock)block andNetwork:(JNSocialDownloadNetwork)network;


- (instancetype)initWithInfoBlock:(SocialDownloadUserInfoBlock)block andNetwork:(JNSocialDownloadNetwork)network;
- (instancetype)initWithImageBlock:(SocialDownloadImageBlock)block andNetwork:(JNSocialDownloadNetwork)network;

@end





@implementation RequestConfiguration

- (instancetype)initWithInfoBlock:(SocialDownloadUserInfoBlock)block andNetwork:(JNSocialDownloadNetwork)network
{
    self = [super init];
    if (self) {
        _network = network;
        _infoBlock = block;
    }
    return self;
}

- (instancetype)initWithImageBlock:(SocialDownloadImageBlock)block andNetwork:(JNSocialDownloadNetwork)network
{
    self = [super init];
    if (self) {
        _network = network;
        _imageBlock = block;
    }
    return self;
}

+ (instancetype)requestConfigurationWithInfoBlock:(SocialDownloadUserInfoBlock)block andNetwork:(JNSocialDownloadNetwork)network
{
    return [[RequestConfiguration alloc] initWithInfoBlock:block andNetwork:network];
}
+ (instancetype)requestConfigurationWithImageBlock:(SocialDownloadImageBlock)block andNetwork:(JNSocialDownloadNetwork)network
{
    return [[RequestConfiguration alloc] initWithImageBlock:block andNetwork:network];
}

@end








#pragma mark - JNSocialDownload

typedef NS_ENUM(NSInteger, SocialDownloadType) {
    SocialDownloadTypeAvatar = 0,
    SocialDownloadTypeCover,
    SocialDownloadTypeInformation
};


const NSInteger JNSocialDownloadNoAPPID = 1;
const NSInteger JNSocialDownloadNoPermissions = 2;
const NSInteger JNSocialDownloadNoAccount = 3;



@interface JNSocialDownload ()

@property (nonatomic) ACAccountStore * accountsStore;

@end


@implementation JNSocialDownload


- (instancetype)initWithAppID:(NSString *)appid
{
    self = [self init];
    if (self) {
        _appID = appid;
    }
    return self;
}

#pragma mark Accessors

- (ACAccountStore *)accountsStore
{
    if (!_accountsStore)
        _accountsStore = [[ACAccountStore alloc]init];
    
    return _accountsStore;
}



#pragma mark - Public API

- (void)downloadAvatar:(SocialDownloadImageBlock)completion forNetwork:(JNSocialDownloadNetwork)network
{
    
    RequestConfiguration * request = [RequestConfiguration requestConfigurationWithImageBlock:completion andNetwork:network];
    
    [self selectSocialIOS:SocialDownloadTypeAvatar request:request];
}
- (void)downloadCover:(SocialDownloadImageBlock)completion forNetwork:(JNSocialDownloadNetwork)network
{
    
    RequestConfiguration * request = [RequestConfiguration requestConfigurationWithImageBlock:completion andNetwork:network];
    
    [self selectSocialIOS:SocialDownloadTypeCover request:request];
}
- (void)downloadInformation:(SocialDownloadUserInfoBlock)completion forNetwork:(JNSocialDownloadNetwork)network
{
    
    RequestConfiguration * request = [RequestConfiguration requestConfigurationWithInfoBlock:completion andNetwork:network];
    
    [self selectSocialIOS:SocialDownloadTypeInformation request:request];
}




#pragma mark - Social Logic

- (void)selectSocialIOS:(SocialDownloadType)downloadType request:(RequestConfiguration *)request
{
    // Social Logic:
    
    // Get ME
    // Get picture with my userID
    
    NSString * SocialAPPID;
    
    if (self.appID)
        SocialAPPID = self.appID;
    else
        SocialAPPID = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"SocialAppID"];
    
    if (!SocialAPPID)
    {
        if (downloadType == SocialDownloadTypeAvatar ||
            downloadType == SocialDownloadTypeCover)
        {
            
            SocialDownloadImageBlock block = request.imageBlock;
            
            block(nil,[NSError errorWithDomain:@"JNSocialDownload" code:JNSocialDownloadNoAPPID userInfo:@{NSLocalizedDescriptionKey:@"No app ID configured"}]);
            
        }
        else if (downloadType == SocialDownloadTypeInformation)
        {
            SocialDownloadUserInfoBlock block = request.infoBlock;
            
            block(nil,[NSError errorWithDomain:@"JNSocialDownload" code:JNSocialDownloadNoAPPID userInfo:@{NSLocalizedDescriptionKey:@"No app ID configured"}]);
        }
        
        return;
    }
    
    
    ACAccountType * accountType;
    NSDictionary * socialOptions;
    
    if (request.network==JNSocialDownloadNetworkFacebook)
    {
            accountType = [self.accountsStore accountTypeWithAccountTypeIdentifier:ACAccountTypeIdentifierFacebook];
            socialOptions = @{ACFacebookAppIdKey:SocialAPPID,
                              ACFacebookPermissionsKey:@[@"user_birthday"]};
        
    }
    else if (request.network == JNSocialDownloadNetworkTwitter)
    {
            accountType = [self.accountsStore accountTypeWithAccountTypeIdentifier:ACAccountTypeIdentifierTwitter];
    }
    
    
    
    [self.accountsStore requestAccessToAccountsWithType:accountType options:socialOptions completion:^(BOOL granted, NSError *error) {
        
        if (error){
            
         
            if (downloadType == SocialDownloadTypeAvatar ||
                downloadType == SocialDownloadTypeCover)
            {
                
                SocialDownloadImageBlock block = request.imageBlock;
                
                block(nil,[NSError errorWithDomain:@"JNSocialDownload" code:JNSocialDownloadNoAccount userInfo:@{NSLocalizedDescriptionKey:@"No Social Account configured"}]);
                
            }
            else if (downloadType == SocialDownloadTypeInformation)
            {
                SocialDownloadUserInfoBlock block = request.infoBlock;
                
                block(nil,[NSError errorWithDomain:@"JNSocialDownload" code:JNSocialDownloadNoAccount userInfo:@{NSLocalizedDescriptionKey:@"No Social Account configured"}]);
            }
            
                
            return ;
        }
        
        if (granted)
        {
            
            [self requestMe:downloadType request:request];
            
        }
        else
        {
            
            if (downloadType == SocialDownloadTypeAvatar ||
                downloadType == SocialDownloadTypeCover)
            {
                
                SocialDownloadImageBlock block =request.imageBlock;
                
                block(nil,[NSError errorWithDomain:@"JNSocialDownload" code:JNSocialDownloadNoPermissions userInfo:@{NSLocalizedDescriptionKey:@"Social permission not granted"}]);
                
            }
            else if (downloadType == SocialDownloadTypeInformation)
            {
                SocialDownloadUserInfoBlock block = request.infoBlock;
                
                block(nil,[NSError errorWithDomain:@"JNSocialDownload" code:JNSocialDownloadNoPermissions userInfo:@{NSLocalizedDescriptionKey:@"Social permission not granted"}]);
            }
            
        }
        
 
 }];
 
 }

- (void)requestMe:(SocialDownloadType)downloadType request:(RequestConfiguration *)request
{
    
    
    SLRequest *merequest = [self requestForNetwork:request.network
                                          dataType:SocialDownloadTypeInformation
                                          userData:nil];
    
    
    __weak typeof(self)weakSelf = self;
    
    [merequest performRequestWithHandler:^(NSData *responseData, NSHTTPURLResponse *urlResponse, NSError *error) {
        
        
        if (!error){
            
            NSError *jsonParsingError = nil;
            NSDictionary * userData = [NSJSONSerialization JSONObjectWithData:responseData options:0 error:&jsonParsingError];
            
            
            if (downloadType == SocialDownloadTypeAvatar)
            {
                [weakSelf requestAvatarFromUser:userData request:request];
            }
            else if (downloadType == SocialDownloadTypeCover)
            {
                [weakSelf requestCoverFromUser:userData request:request];
            }
            else if (downloadType == SocialDownloadTypeInformation)
            {
                SocialDownloadUserInfoBlock block =request.infoBlock;
                
                block(userData,nil);
            }
            
            
            
        }
        
    }];

}

- (void)requestAvatarFromUser:(NSDictionary *)userData request:(RequestConfiguration *)request
{
    
    SLRequest * avatarRequest  = [self requestForNetwork:request.network
                                          dataType:SocialDownloadTypeAvatar
                                          userData:userData];
    
    
    __weak typeof(self)weakSelf = self;


    [avatarRequest performRequestWithHandler:^(NSData *responseData, NSHTTPURLResponse *urlResponse, NSError *error) {
        
         [weakSelf processImageData:responseData request:request];
        
    }];
}


- (void)requestCoverFromUser:(NSDictionary *)userData request:(RequestConfiguration *)request
{
    
    SLRequest * coverRequest  = [self requestForNetwork:request.network
                                          dataType:SocialDownloadTypeCover
                                          userData:userData];
    
    
    __weak typeof(self)weakSelf = self;

    
    [coverRequest performRequestWithHandler:^(NSData *responseData, NSHTTPURLResponse *urlResponse, NSError *error) {
        
        if (request.network == JNSocialDownloadNetworkFacebook)
        {
            
            NSDictionary * response = [NSJSONSerialization JSONObjectWithData:responseData options:0 error:nil];
            
            BOOL anyError = NO;
            
            if (response)
            {
                
                NSDictionary * cover = response[@"cover"];
                if (cover){
                    NSString * url = cover[@"source"];
                    
                    SLRequest * request3 =
                    [SLRequest requestForServiceType:SLServiceTypeFacebook
                                       requestMethod:SLRequestMethodGET
                                                 URL:[NSURL URLWithString:url]
                                          parameters:nil];
                    
                    [request3 performRequestWithHandler:^(NSData *responseData, NSHTTPURLResponse *urlResponse, NSError *error) {
                        
                         [weakSelf processImageData:responseData request:request];
                        
                        
                    }];
                    
                }
                else
                {
                    anyError = YES;
                }
            }
            else
            {
                anyError = YES;
            }
            
            if (anyError)
            {
                ////TODO: throw an error
            }
            
            
        }
        
        else if (request.network == JNSocialDownloadNetworkTwitter)
        {
            [weakSelf processImageData:responseData request:request];
        }
        
        
    }];

}

- (void)processImageData:(NSData *)imageData request:(RequestConfiguration *)request
{
    if (imageData.length > 100)
    {
        UIImage * image = [UIImage imageWithData:imageData];
        
        SocialDownloadImageBlock block = request.imageBlock;
        
        block(image,nil);
    }
}

#pragma mark Request preparation
- (SLRequest *)requestForNetwork:(JNSocialDownloadNetwork)network
                        dataType:(SocialDownloadType)type
                        userData:(NSDictionary *)userData
{
    
    if (network == JNSocialDownloadNetworkFacebook)
    {

        SLRequest * request;
        
        
        if (type == SocialDownloadTypeInformation)
        {
            
            NSURL * meurl = [NSURL URLWithString:@"https://graph.facebook.com/me"];
            
            
            request = [SLRequest requestForServiceType:SLServiceTypeFacebook
                                         requestMethod:SLRequestMethodGET
                                                   URL:meurl
                                            parameters:nil];
            
            
        }
        else if (type == SocialDownloadTypeAvatar)
        {
            
            NSString * SocialID = userData[@"id"];
            NSString * imageURL = [NSString stringWithFormat:@"https://graph.facebook.com/v2.1/%@/picture",SocialID];
            
            
            request = [SLRequest requestForServiceType:SLServiceTypeFacebook
                                         requestMethod:SLRequestMethodGET
                                                   URL:[NSURL URLWithString:imageURL]
                                            parameters:@{@"height":@"512",
                                                         @"width":@"512"}];
            
        }
        else if (type == SocialDownloadTypeCover)
        {
            
            NSString * SocialID = userData[@"id"];
            NSString * coverURL = [NSString stringWithFormat:@"https://graph.facebook.com/v2.1/%@",SocialID];
            
            
            
            request = [SLRequest requestForServiceType:SLServiceTypeFacebook
                                         requestMethod:SLRequestMethodGET
                                                   URL:[NSURL URLWithString:coverURL]
                                            parameters:@{@"fields":@"cover"}];
            
        }
        
        ACAccountType * accountType = [self.accountsStore accountTypeWithAccountTypeIdentifier:ACAccountTypeIdentifierFacebook];
        ACAccount * account = [[self.accountsStore accountsWithAccountType:accountType] lastObject];
        request.account = account;
        
        
        return request;
        
    }
    
    
    else if (network == JNSocialDownloadNetworkTwitter)
    {
        
        SLRequest * request;
        
        
        if (type == SocialDownloadTypeInformation)
        {
            
            NSURL * meurl = [NSURL URLWithString:@"https://api.twitter.com/1.1/account/verify_credentials.json"];
            
            
            request = [SLRequest requestForServiceType:SLServiceTypeTwitter
                                         requestMethod:SLRequestMethodGET
                                                   URL:meurl
                                            parameters:nil];
            
        }
        else if (type == SocialDownloadTypeAvatar)
        {
            NSString * imageURL = userData[@"profile_image_url"];
            
            
            request = [SLRequest requestForServiceType:SLServiceTypeTwitter
                                         requestMethod:SLRequestMethodGET
                                                   URL:[NSURL URLWithString:imageURL]
                                            parameters:nil];
        }
        else if (type == SocialDownloadTypeCover)
        {
            
            NSString * imageURL = userData[@"profile_background_image_url"];
            
            request = [SLRequest requestForServiceType:SLServiceTypeTwitter
                                         requestMethod:SLRequestMethodGET
                                                   URL:[NSURL URLWithString:imageURL]
                                            parameters:nil];

            
        }
        
        ACAccountType * accountType = [self.accountsStore accountTypeWithAccountTypeIdentifier:ACAccountTypeIdentifierTwitter];
        ACAccount * account = [[self.accountsStore accountsWithAccountType:accountType] lastObject];
        request.account = account;
        
        
        return request;
    }
    
    return nil;
}


@end

File: JNSocialDownload-Swift/JNSocialDownload.swift
//
//  JNSocialDownload.swift
//  JNSocialDownloadDemo
//
//  Created by Joao Nunes on 04/02/15.
//  Copyright (c) 2015 joao. All rights reserved.
//

import UIKit
import Accounts
import Social
import Foundation


private enum SocialDownloadType
{
    case Avatar
    case Cover
    case Information
}

enum JNSocialDownloadNetwork
{
    case Facebook
    case Twitter
}

let JNSocialDownloadNoAPPID = 1
let JNSocialDownloadNoPermissions = 2
let JNSocialDownloadNoAccount = 3



typealias SocialDownloadImageClosure = (image : UIImage?, error: NSError?) -> ()
typealias SocialDownloadInformationClosure = (userInfo : NSDictionary?, error: NSError?) -> ()


private class RequestConfiguration
{
    let network:JNSocialDownloadNetwork
    
    let imageClosure:SocialDownloadImageClosure?
    let infoClosure:SocialDownloadInformationClosure?

    
    init(infoClosure:SocialDownloadInformationClosure, network:JNSocialDownloadNetwork)
    {
        self.network = network
        self.infoClosure = infoClosure
    }
    init(imageClosure:SocialDownloadImageClosure, network:JNSocialDownloadNetwork)
    {
        self.network = network
        self.imageClosure = imageClosure
    }
}


class JNSocialDownload: NSObject {
   
    
    var appID:NSString?

    lazy var accountStore: ACAccountStore =
    {
        var tempAccountStore: ACAccountStore = ACAccountStore()
        
        return tempAccountStore
    }()
    

    convenience init(appID:NSString)
    {
        self.init()
        self.appID = appID
    }
    
    
    
    func downloadInformation(completionHandler: SocialDownloadInformationClosure, network:JNSocialDownloadNetwork) -> Void
    {
        let request = RequestConfiguration(infoClosure: completionHandler,network: network)
        
        self.selectSocialIOS(.Information, request: request)
    }
    func downloadAvatar(completionHandler: SocialDownloadImageClosure, network:JNSocialDownloadNetwork) -> Void
    {
        let request = RequestConfiguration(imageClosure: completionHandler,network: network)
        
        self.selectSocialIOS(.Avatar, request: request)
    }
    
    func downloadCover(completionHandler: SocialDownloadImageClosure, network:JNSocialDownloadNetwork) -> Void
    {
        let request = RequestConfiguration(imageClosure: completionHandler,network: network)
        
        self.selectSocialIOS(.Cover, request: request)
    }
    
}

// MARK: Logic
extension JNSocialDownload
{
    
    private func selectSocialIOS(downloadType: SocialDownloadType, request:RequestConfiguration)
    {
     
        var socialAPPID :NSString?
        
        if let appID = self.appID {
            socialAPPID = appID
        }
        else
        {
            socialAPPID = NSBundle.mainBundle().objectForInfoDictionaryKey("SocialAppID") as NSString?
        }
        
        
        if (socialAPPID == nil)
        {
            
            if (downloadType == .Avatar ||
                downloadType == .Cover)
            {
                
                
                if let closure = request.imageClosure
                {
                    closure(image: nil, error: NSError(
                        domain: "JNSocialDownload",
                        code: JNSocialDownloadNoAPPID,
                        userInfo: [NSLocalizedDescriptionKey:"No app ID configured"]))
                }
                
                
            }
            else if (downloadType == .Information)
            {
                if let closure = request.infoClosure
                {
                    closure(userInfo: nil, error: NSError(
                        domain: "JNSocialDownload",
                        code: JNSocialDownloadNoAPPID,
                        userInfo: [NSLocalizedDescriptionKey:"No app ID configured"]))
                }
            }
            
            return
        }
        
        var accountType :ACAccountType!
        var socialOptions:Dictionary<String,AnyObject>? = nil
        
        if (request.network == .Facebook)
        {
            accountType = self.accountStore.accountTypeWithAccountTypeIdentifier(ACAccountTypeIdentifierFacebook)
            socialOptions = [ACFacebookAppIdKey:socialAPPID!,
                ACFacebookPermissionsKey:["user_birthday"]]
        }
        else if (request.network == .Twitter)
        {
            accountType = self.accountStore.accountTypeWithAccountTypeIdentifier(ACAccountTypeIdentifierTwitter)
        }
        
        
        
        self.accountStore .requestAccessToAccountsWithType(accountType, options: socialOptions, completion: { (granted, error) -> Void in
            
            if (error != nil)
            {
                
                
                if (downloadType == .Avatar ||
                    downloadType == .Cover)
                {
                    
                    
                    if let closure = request.imageClosure
                    {
                        closure(image: nil, error: NSError(
                            domain: "JNSocialDownload",
                            code: JNSocialDownloadNoAccount,
                            userInfo: [NSLocalizedDescriptionKey:"No Social Account configured"]))
                    }
                    
                    
                }
                else if (downloadType == .Information)
                {
                    if let closure = request.infoClosure
                    {
                        closure(userInfo: nil, error: NSError(
                            domain: "JNSocialDownload",
                            code: JNSocialDownloadNoAccount,
                            userInfo: [NSLocalizedDescriptionKey:"No Social Account configured"]))
                    }
                }
                
                
                return
            }
            
            
            if (granted)
            {
                self.requestMe(downloadType, request: request)
            }
            else
                
            {
                if (downloadType == .Avatar ||
                    downloadType == .Cover)
                {
                    
                    
                    if let closure = request.imageClosure
                    {
                        closure(image: nil, error: NSError(
                            domain: "JNSocialDownload",
                            code: JNSocialDownloadNoPermissions,
                            userInfo: [NSLocalizedDescriptionKey:"Social permission not granted"]))
                    }
                    
                    
                }
                else if (downloadType == .Information)
                {
                    if let closure = request.infoClosure
                    {
                        closure(userInfo: nil, error: NSError(
                            domain: "JNSocialDownload",
                            code: JNSocialDownloadNoPermissions,
                            userInfo: [NSLocalizedDescriptionKey:"Social permission not granted"]))
                    }
                }
                
            }
            
            
        })
        

       
    }
    
    
    
    private func requestMe(downloadType: SocialDownloadType, request:RequestConfiguration)
    {
        
        
        let merequest = self.requestFor(
            request.network,
            dataType: .Information,
            userData: nil)
        
        
        merequest?.performRequestWithHandler( { [weak self,request] (responseData, urlResponse, error) -> Void in
            
            let userData = NSJSONSerialization.JSONObjectWithData(responseData!, options: NSJSONReadingOptions.allZeros, error: nil) as Dictionary<String,AnyObject>
            
            if (downloadType == .Avatar)
            {
                self?.requestAvatarFrom(userData, request: request)
            }
            else if (downloadType == .Cover)
            {
                self?.requestCoverFrom(userData, request: request)
            }
            else if (downloadType == .Information)
            {
                let closure = request.infoClosure!
                
                closure(userInfo: userData , error: nil)
                
            }
            
        })
        
        
        
    }
    
    
    private func requestAvatarFrom(userData:Dictionary<String,AnyObject>, request:RequestConfiguration)
    {
        
        let avatarRequest = self.requestFor(
            request.network,
            dataType: .Avatar,
            userData: userData)

        // (NSData!, NSHTTPURLResponse!, NSError!) -> Void
        avatarRequest?.performRequestWithHandler( { (responseData: NSData!, urlResponse:NSHTTPURLResponse!, error:NSError!) -> Void in
            
            self.processImage(responseData, request: request)
            
        })
        
    }
    private func requestCoverFrom(userData:Dictionary<String,AnyObject>, request:RequestConfiguration)
    {
        let coverRequest = self.requestFor(
            request.network,
            dataType: .Cover,
            userData: userData)
        
        
        coverRequest?.performRequestWithHandler( {( responseData, urlResponse, error) -> Void in
            
            if (request.network == .Facebook)
            {
                let response = NSJSONSerialization.JSONObjectWithData(responseData, options: NSJSONReadingOptions.allZeros, error: nil) as Dictionary<String,AnyObject>?
                
                let anyError = false
                
                if (response != nil)
                {
                    let cover = response!["cover"] as Dictionary<String,AnyObject>?
                    if (cover != nil)
                    {
                        
                        let url = cover!["source"] as String
                        
                        let request3 = SLRequest(forServiceType: SLServiceTypeFacebook, requestMethod: .GET, URL: NSURL(string: url), parameters: nil)
                        
                        
                        request3.performRequestWithHandler( {(responseData, urlResponse, error) -> Void in
                            
                            self.processImage(responseData, request: request)
                            
                        })

                        
                    }
                }
                
                
            }
            else if (request.network == .Twitter)
            {
                self.processImage(responseData, request: request)
            }
            
        })
        
    }
    
    
    
    private func processImage(data:NSData, request:RequestConfiguration)
    {
        if (data.length > 100)
        {
            let image = UIImage(data: data)
            
            let closure = request.imageClosure!
            
            closure(image: image,error: nil)
        }
    }
    
   // MARK: Request preparation
    private func requestFor(
        network: JNSocialDownloadNetwork,
        dataType:SocialDownloadType,
        userData:Dictionary<String,AnyObject>?) -> SLRequest?
    {
        if (network == .Facebook)
        {
            var request:SLRequest?
            
            if (dataType == .Information)
            {
                let meurl = NSURL(string: "https://graph.facebook.com/me")
                
                request = SLRequest(
                    forServiceType: SLServiceTypeFacebook,
                    requestMethod: .GET,
                    URL: meurl,
                    parameters: nil)
                
            }
            else if (dataType == .Avatar)
            {
                let socialID = userData!["id"] as String
                let imageURL = "https://graph.facebook.com/v2.1/\(socialID)/picture"
                
                request = SLRequest(
                    forServiceType: SLServiceTypeFacebook,
                    requestMethod: .GET,
                    URL: NSURL(string: imageURL),
                    parameters: ["height":"512","width":"512"])
                
            }
            else if (dataType == .Cover)
            {
                let socialID = userData!["id"] as String
                let coverURL = "https://graph.facebook.com/v2.1/\(socialID)"
                
                request = SLRequest(
                    forServiceType: SLServiceTypeFacebook,
                    requestMethod: .GET,
                    URL: NSURL(string: coverURL),
                    parameters: ["fields":"cover"])
            }
            
            let accountType = self.accountStore.accountTypeWithAccountTypeIdentifier(ACAccountTypeIdentifierFacebook)
            let account = self.accountStore.accountsWithAccountType(accountType).last! as ACAccount
            
            if (request != nil){
                request!.account = account
            }
            
            return request
            
        }
        else if (network == .Twitter)
        {
            
            var request:SLRequest?
            
            if (dataType == .Information)
            {
                let meurl = NSURL(string: "https://api.twitter.com/1.1/account/verify_credentials.json")
                
                request = SLRequest(
                    forServiceType: SLServiceTypeTwitter,
                    requestMethod: .GET,
                    URL: meurl,
                    parameters: nil)
                
            }
            else if (dataType == .Avatar)
            {
                let imageURL = userData!["profile_image_url"] as String
                
                request = SLRequest(
                    forServiceType: SLServiceTypeTwitter,
                    requestMethod: .GET,
                    URL: NSURL(string: imageURL),
                    parameters: nil)
                
            }
            else if (dataType == .Cover)
            {
                let imageURL = userData!["profile_background_image_url"] as String
                
                request = SLRequest(
                    forServiceType: SLServiceTypeTwitter,
                    requestMethod: .GET,
                    URL: NSURL(string: imageURL),
                    parameters: nil)
            }
            
            let accountType = self.accountStore.accountTypeWithAccountTypeIdentifier(ACAccountTypeIdentifierTwitter)
            let account = self.accountStore.accountsWithAccountType(accountType).last! as ACAccount
            
            if (request != nil){
                request!.account = account
            }
            
            return request

        }
        
        return nil
        
    }
    

}
